### DISCUSSION TITLE
Page structure allowing new features

### DISCUSSION THREAD

**Discussion by Joduai** [danpros/htmly#825] [Ideas]
For quite some time I've been thinking about starting this topic.  
Current page structure, and the reasons it stays that way have been already mentioned in #718  
But with new requests, and ideas adding even small things becomes problematic.  
Every possible page option can't be stored in filename, and processing whole file to get the option stored in custom` <!-- tag --> ` is tedious.  

**How about moving page's options to new file attached to its parent?**  
For example by generating page ID, saving this ID in page name as an additional parameter, and use this ID for option's filename.
This ID could be even the **page's creation date**, thus we would get
```
2024-10-14-19-00-01_tag1_tag2_page-name.md
2024-10-14-19-00-01_options.md (or any other extension)
```
this way all content stays without changes in future updates, and we get new possibilities with additional page options.
For backwards compatibility, and to prevent data loss, all tags stored in posts files should still be there.
Just to mention `<!--t title t--> <!--d description d--> <!--tag tags tag--> <!--image page type header-->`
Why? If page's options file becomes corrupted, or has no data at all, htmly would still display all content just like it does now as a fallback, skipping unavailable data from options file.

I'm not sure about the data structure for options file. Probably in json format, maybe xml or a custom one, you name it.
Changing or adding new options to page wouldn't update post file (except currently existing page options like tag, category...) thus it wouldn't have impact on last modification date.

In some use cases data shown on website - like showing blog posts with featured images and excerpts - wouldn't even process the post file, because all data could be taken from options file. Like image/video/quote, excerpt, tags, category, author etcetera.
Even adding an option to pin a page as a featured post on top of the list would be available straight away, although this one option is somehow doable by using tags. But using tags as additional option for missing features creates new problems along solving others.

I asked earlier for custom options for admin panel, similar to adding new fields in the custom settings for using them in templates.  
This would open possibilities to create custom page's options by page admins.   

Features that were asked and would benefit from this I can imagine right now are
- robots meta tags, allowing to index page or archive it
- enabling/disabling comments
- allowing other editors to edit post
- custom fields per page
- assign different template file to a page

Any thoughts about this whole idea?

**Comment by danpros** [comment]
Hello,

Yes we should add per post/page options the core, and because of there is no ID than we can use the slug for the ID, example:

```
// blog post
post_postSlug.json

// page
page_pageSlug.json

// Subpage
subpage_parentSlug.subpageSlug.json
```

For folder we can put it in example, `content/data/options`.

**Reply by Joduai** [reply]
storing page options in 1 directory rather than along with posts/pages files is a good idea. Less file operations after changing post's/page's title, slug etc.
While having on mind a bit bigger blogs or websites maybe splitting options into individual dirs for content/data/options/posts/ and /pages/ would turn out a bit easier to use later? As for now can't imagine use case for it. Maybe for displaying on website a list of pages or posts along with some of their parameters without the hassle of using exceptions, in some kind of a sitemap.

Using slug as ID seems also a good point. Page's creation date is just unchangeable, and therefore unique. But afair there can't be 2 same slugs, as new file gets a "timestamp" in slug when the same slug already exists.

**Reply by danpros** [reply]
The page title, descriptions etc. should not added to the options file. This options file only stores additional data especially for per post/page configurations or custom fields. For the data structures, we need to think about which is best. 

Managing the contents in htmly without using the dashboard is very easy. For example, we have a user "admin" and there are several posts (url `example.com/author/admin`), and we want to change the ownership to eg, `new-user`, simply rename the `content/admin` to `content/new-user`, clear cache and its done.

Easy to change the domain or hosting etc. so we should maintain this ease.

**Reply by Joduai** [reply]
I'm fully aware of htmly's headless content management functionality, and I'm very fond of it. That's why I insisted on leaving current page/post file name and structure, while allowing for further development with new functions.
Although I can't imagine how in the future, if ever, this arrangement could change. 

Now I'm curious why not storing all useful data in options files, along with title, descriptions and so on?
It's only a small chunk of additional data, that won't have any visible impact on performance. 
While still making a bit easier to operate on all attributes stored in 1 place, depending of course on chosen data structure. 
I am personally more used to operate on xml structures on shared webhostings rather than with data stored in json. 

From the other hand  maybe storing page attributes in external file is a faulty approach?
Is there any crucial reason why not transforming text file with content into structurized json or xml format? 
I understand the KISS approach in this project, as you mentioned in some issues. 
Loving it fully, but one day inevitably this becomes a bottleneck. What then?
One of the most important cons that such change would complicate headless content creation.
But aren't such creators used to many types of tags typed by hand, so that using in document markup like _<title>, <tags>_ etc. is somehow natural. Especially for those using markdown, as it demands even frequently using html tags.

-----

As it goes to easily changing domains matter, afair I pointed this out. Switching domains for CMS is often pretty easy, especially in file based systems.  
The problem stays with absolute urls (pages, media files etc) in content, that aren't easy to find and replace. 
That's why I asked in one of QA issues if there's an easy way to switch from absolute to relative urls. 
But this is a different matter, so let's leave it for now.

**Reply by taufik-nurrohman** [reply]
> It's only a small chunk of additional data, that won't have any visible impact on performance.

But slow to copy-paste.
 https://www.quora.com/Why-does-copying-multiple-files-take-longer-time-than-copying-a-single-file-of-same-size

**Reply by Joduai** [reply]
> But slow to copy-paste.  

what exactly does copying large amount of small files - being completely out of scope in this thread - has to do with storing couple additional bytes in text files, that would exist anyway?

**Reply by taufik-nurrohman** [reply]
Export-import large amount of small size data will be problematic in the future.

**Reply by Joduai** [reply]
ohh, backups are a way different story
didn't dig into htmly backup functions, but all content files are zipped right away
there's no import function, but those who send a bunch of files through (s)ftp instead of unarchiving them on destination place, surely have plenty of spare time. Or do it on purpose ;)

Nonetheless this is still way far from this thread. I already mentioned 2 possible ways. Missing page options files could be just recreated basing on default options set for pages. Or we store everything in 1 file but advance from plain txt to structurized markup. I winder if there's any other possible solution?

**Reply by danpros** [reply]
> Now I'm curious why not storing all useful data in options files, along with title, descriptions and so on?

Because those data (title, description, tags display, post type data, post content) already available in the main post file, the data accessible in `$post` so we should not add this data elsewhere. If in WP its like putting some of the wp_posts data to wp_options.

> Is there any crucial reason why not transforming text file with content into structurized json or xml format?

Due to preference and its convenience. Markdown is widely used so I believe there is no problem using it.

> Loving it fully, but one day inevitably this becomes a bottleneck. What then?

With the current data structure it is unlikely that we will experience a bottleneck.

Projects/services/products will focus on a specific niche/utility, and htmly niche is a simple flat-file blog, so the features that are prioritized to be added are features that are usually found on a simple blog in general.

For options and custom fields, we should use 1 option/config file per each post/page. 

And to access the options or custom fields, perhaps:

```php
$myField = get_field('field', $post->slug); // Return all field data for this post
$myOption = get_option('option', $post->slug); // Return all options data for this post

$post->field = $myField;
$post->option = $myOption;
```

We will have `$post->field['my_custom_field'];` or `$post->option['disable_comment'];` and so on.
